### DISCUSSION TITLE
Performance deficit between using evaluation() and multiple custom implementations

### DISCUSSION THREAD

**Discussion by bowentkruse** [mindee/doctr#1410] [Q&A]
When I run --test-only on my test dataset for a custom trained model using the sar_resnet31 architecture on a custom dataset, I get 
a partial and complete match score in the mid-to-low 90s. 

When loading my checkpoint using the recommended method, and running evaluation on my dataset's test set, exact match and partial match is 0%. 

How I'm evaluating on my datasets test set:
`python train_pytorch.py sar_resnet31 --test-only --resume sar_resnet31_20231208-145408.pt --val_path=stencils-1/test --vocab english --pretrained --input_size 48 -b 64`
Output: `Validation loss: 0.0948886 (Exact: 92.81% | Partial: 92.81%)`

I've loaded the model in two ways:
First, as listed in [these docs](https://mindee.github.io/doctr/using_doctr/custom_models_training.html):
```
from doctr.io import DocumentFile
from doctr.models import ocr_predictor

DET_CKPT = "file:///path/to/detection/model/db_resnet50_XXXX/weights"
REC_CKPT = "file://path/to/recognition/model/sar_resnet31_XXXX/weights"

model = ocr_predictor(det_arch='db_resnet50', reco_arch='sar_resnet31',pretrained=True)
model.det_predictor.model.load_weights(DET_CKPT)
model.det_predictor.model.postprocessor.unclip_ratio = 2
model.reco_predictor.model.load_weights(REC_CKPT)

if __name__ == "__main__":
    # Image loading
    doc = DocumentFile.from_images("./path/to/image")
    # Models inference
    result = model(doc)
    # Display the detection and recognition results on the image
    result.show(doc)
```

Second, by copying how the model is loaded and utilized in the `evaluate()` method in train_pytorch.py in the references directory of this repo:

```
from torch.utils.data import DataLoader
from torchvision.transforms.v2 import Compose, Normalize
import torch
from torchvision import transforms
from doctr.datasets import VOCABS
from doctr.models import recognition
from doctr.io import read_img_as_tensor, tensor_from_numpy
from typing import List
from doctr import transforms as T
import numpy as np
import os
import json

transform_pipeline = Compose([
    T.Resize((48, 48 * 4), preserve_aspect_ratio=True),
    transforms.Normalize(mean=(0.694, 0.695, 0.693), std=(0.299, 0.296, 0.301))
])

# Load model architecture and state of given checkpoint
def load_model(arch, vocab, checkpoint):
    model = recognition.__dict__[arch](pretrained=False, pretrained_backbone=False, vocab=VOCABS[vocab])
    model_checkpoint = torch.load(checkpoint, map_location='cpu')
    model.load_state_dict(model_checkpoint)
    if torch.cuda.is_available():
        # Map to single GPU
        torch.cuda.set_device(0)
        model = model.cuda()

    return model

def infer(batch, model):
    model.eval()
    batch = batch.cuda()
    with torch.no_grad():
        predictions = model(batch)
    return predictions

def preprocess_images(image_paths:List) -> torch.Tensor:
    processed_images = []

    for image_path in image_paths:
        
        img = (tensor_from_numpy(image_path, dtype=torch.float32)
               if isinstance(image_path, np.ndarray)
               else read_img_as_tensor(image_path, dtype=torch.float32)
               )
        img = transform_pipeline(img)
        processed_images.append(img)

    processed_images = torch.stack(processed_images)

    return processed_images

def read_coco_json(json_filename):
    with open(json_filename, 'r') as file:
        data = json.load(file)
    return data
# Path to my test directory
test_dir_path = 'OCRDatasets/stencils-1/test'
# Load the labels for each image to compare inference to ground truth
labels = read_coco_json(os.path.join(test_dir_path, 'labels.json'))
image_fns = list(labels.keys())
# Get full paths of each filename
image_paths = [os.path.join(test_dir_path, f"images/{fn}") for fn in image_fns]
# Preprocess all images in the direct directory into a batch
batch = preprocess_images(image_paths)
# Load the model how the code in train_pytorch.py does in the references directory
model = load_model('sar_resnet31', 'ascii_letters', f'Weights/sar_resnet31_20231212-132905.pt')
# infer on loaded batch
predictions = infer(batch, model)
```
Results from both methods produce predicted strings that are of the right vocabulary and have at most 1-2 of 4 letters correct. 
While debugging I've found that using different data loading methods changes the output sof the model significantly (all still incorrect) so
my thought is that my data handling is not consistent with how train_pytorch.py in the references dir does data handling. 

All thoughts/recommendations are appreciated.

**Comment by bowentkruse** [comment]
Update:

I've also just tried the below method with similar unsatisfactory results

```
import os
import torch
from doctr.models.recognition.predictor import RecognitionPredictor
from doctr.models.preprocessor import PreProcessor
from doctr.models import sar_resnet31
from doctr.datasets import VOCABS
from doctr.io import DocumentFile
import json

reco_model = sar_resnet31(pretrained=False, pretrained_backbone=False, vocab=VOCABS['ascii_letters'])

reco_params = torch.load("Weights/sar_resnet31_20231212-132905.pt", map_location="cpu")

reco_model.load_state_dict(reco_params)

reco_predictor = RecognitionPredictor(PreProcessor((48, 48 * 4), preserve_aspect_ratio=True, batch_size=16, mean=(0.694, 0.695, 0.693), std=(0.299, 0.296, 0.301)), reco_model)

json_filename = 'OCRDatasets/stencils-1/test/labels.json'
with open(json_filename, 'r') as file:
    data = json.load(file)

for image_fn in list(data.keys()):

    test_path = os.path.join("OCRDatasets/stencils-1/test/images", image_fn)

    doc = DocumentFile.from_images(test_path)

    result = reco_predictor(doc)

    print(result)
```

**Comment by felixT2K** [comment]
Hi @bowenkruse 👋🏼,
You are right it looks like there is a bug with the LSTMCell's inside the decoder part of the model in both backends (TF and PT) (feel free to open a bug ticket).

**Reply by bowentkruse** [reply]
Hey @felixT2K, thank you so much for the quick reply. I've created a [bug report issue here](https://github.com/mindee/doctr/issues/1411). Please let me know if there is any additional information I can offer to assist in debugging.

**Reply by felixdittrich92** [reply]
Hey @bowenkruse thanks for the issue we will have a look (i think next year ^^).
For the meanwhile you can try for example the `parseq` implementation :) (which is faster, smaller and more accure)

**Reply by bowentkruse** [reply]
You're a mind reader @felixdittrich92, there's a `parseq` model training on my computer as you posted that. Thanks again.

**Reply by felixdittrich92** [reply]
Let me know if you have questions or some trouble ^^
